### ISSUE TITLE
Getting passwords is slow?

### ISSUE THREAD

**Issue by dabeeeenster** [doy/rbw#11]
Is this as expected?

```
time rbw get --full "<password name>"  0.26s user 1.84s system 94% cpu 2.224 total
```

This is on a core i7 2020 Macbook Pro

**Comment by doy** [comment]
how many passwords do you have in your vault? on linux, with ~200 passwords in my vault, `rbw get --full` takes ~80ms for me. also, how are you building it? it'll definitely also be a lot slower in debug mode.

**Comment by dabeeeenster** [comment]
I have 530 passwords. I'm building with `cargo install rbw`.

**Comment by jakeswenson** [comment]
I also have around 570 passwords in bitwarden and getting a single password takes around 3.5 seconds

```
time rbw get "<password name>"  0.40s user 3.48s system 97% cpu 3.980 total
```

Not quite sure on the encrypted storage format of the Db yet, but maybe this is a case for an embedded database... I've been watching sled closely (probably way overkill for this project) but there are probably some improvements that could be made here.

**Comment by doy** [comment]
i'd be pretty curious to see some profiling results here before digging into options for a fix. i'd be quite surprised if the issue here is the on-disk format - it seems a lot more likely that it's something like one of the crypto libraries i'm using not having optimized implementations of certain functions on certain platforms.

**Comment by jakeswenson** [comment]
Here is a (screenshot of a) flame graph I generated of `rbw ls`:
![image](https://user-images.githubusercontent.com/1854799/88459399-1bf27e00-ce4a-11ea-8949-6d74948ec39b.png)
80%+ of the time is spent in `ttyname`

(screenshot since github wont let me add the SVG)

I used `cargo-flamegraph` to generate this.

**Comment by jakeswenson** [comment]
Although it might not be the best idea, I was able to speed it up *drastically* by removing all the code in `ttyname` and replacing it with this:
![image](https://user-images.githubusercontent.com/1854799/88459619-9374dd00-ce4b-11ea-9a8a-c52d35e1859c.png)

Now timing `cargo ls`:
```
rbw ls  0.04s user 0.08s system 54% cpu 0.224 total
```

**Comment by jakeswenson** [comment]
Now 92% of the time is spent in `decrypt_cipher` which seems reasonable:
![image](https://user-images.githubusercontent.com/1854799/88459722-607f1900-ce4c-11ea-98ab-8e0f57b7c343.png)

**Comment by jakeswenson** [comment]
Maybe this isn't such a bad idea: https://unix.stackexchange.com/a/450726
Has the added benefit of removing an unsafe block and the only usage of `libc`

**Comment by doy** [comment]
no, this is not actually correct - we are trying to pass the name of the tty to the background agent process, which is entirely independent, so if you start the agent in one terminal, lock it, and then try to unlock it from a second terminal, `/dev/tty` from the perspective of the agent process is not going to be a meaningful thing. i have no idea why `ttyname_r` would be so slow though...

**Comment by doy** [comment]
do you also get similar results if you run something like this? ```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int main(int argc, char *argv[]) {
    char buf[1024];
    int err = ttyname_r(0, buf, 1024);
    if (err != 0) {
        fprintf(stderr, "%s\n", strerror(err));
        exit(1);
    }
    puts(buf);
}
```

**Comment by jakeswenson** [comment]
That's a weird-looking rust program ;P

Nope, that program runs almost instantly:
`clang code.c`

```bash
 time ./a.out
/dev/ttys001
./a.out  0.00s user 0.00s system 74% cpu 0.005 total
```

**Comment by albakham** [comment]
For me, getting password or listing the entries is rather quick.

What I find to be slow is:
after I `rbw edit` an entry, I exit my editor (vim/nano/whatever), and it takes several seconds before I can use my shell again.

I suppose it needs to update some things, but perhaps could it be improved? at least when no changes are made? (because for now, changes or no changes, it's always as slow)

**Comment by stevegore** [comment]
It's a shame this couldn't be reproduced. Getting passwords is painfully slow for me as well. Mind you I'm on a Mac so I appreciate I'm probably not your target audience here.

**Comment by dabeeeenster** [comment]
I'm not a rust dev I'm afraid but happy to help if its easy for me to generate debugging info that would help?

**Comment by JaCoB1123** [comment]
I was just about to write that I'm having the same problem on Manjaro. My vault is quite big with 2130 entries, so I though this might be the reason.
I was using the seemingly outdated rbw-git package though and after upgrading to 1.8.3 from 1.7.1 `ls` and `get` are now much faster. Before upgrading a `get` would take at least 1,5 seconds and sometimes more than 5 seconds. After upgrading it's now consistently at around 0,6 seconds. 

Can anyone else confirm this?

**Comment by domrim** [comment]
> I was just about to write that I'm having the same problem on Manjaro. My vault is quite big with 2130 entries, so I though this might be the reason. I was using the seemingly outdated rbw-git package though and after upgrading to 1.8.3 from 1.7.1 `ls` and `get` are now much faster. Before upgrading a `get` would take at least 1,5 seconds and sometimes more than 5 seconds. After upgrading it's now consistently at around 0,6 seconds.
> 
> Can anyone else confirm this?

nope, i'm also using rbw 1.8.3 on macOS 13.5 (22G74). `rbw ...` takes still about 10s per command :(

**Comment by gdw2** [comment]
Not sure this is related, but on macOS I noticed that `pinentry` (subprocess of `rbw-agent`) is running at 100% cpu all the time -- until I force close it.

**Comment by baleksa** [comment]
> Not sure this is related, but on macOS I noticed that `pinentry` (subprocess of `rbw-agent`) is running at 100% cpu all the time -- until I force close it.

Same here. `Voidlinux x86_64`. Two `pinentry` subprocesses, both of which were 99% CPU reported by `htop`, and two CPU cores were at 99%, but there was no performance drop and fans weren't spinning.
I will provide more details and possibly open an issue when I get time to examine it better.

**Comment by doy** [comment]
i think the issue with `pinentry` using an entire cpu is different from the issue that people were seeing earlier - i would like to fix that one too though, so let me know if you manage to reproduce it.

**Comment by pchampio** [comment]
Hello, same issue with `pinentry` taking 99% of CPU.
After playing around, it seems that it is a `pinentry-curses` issue (super old issue https://dev.gnupg.org/T814 related?).   
Using `pinentry-tty` fixes the issue.
```
rbw config set pinentry "/usr/bin/pinentry-tty"
```

**Comment by sedlund** [comment]
when trying to `rbw get` 

the `rbw-agent` proccess strace is full of this

> sudo strace -fp $(pgrep rbw-agent) 2> out

> rg "FUTEX|unavail" out
```strace
163971:[pid 2035804] futex(0x7ffff7cb7d00, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>
163983:[pid 2035817] futex(0x7ffff7cb7d00, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
163986:[pid 2035819] futex(0x7ffff607e640, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>
163994:[pid 2035804] futex(0x7ffff607e640, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
164001:[pid 2035804] <... accept4 resumed>0x7fffffff6c60, [110], SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
164004:[pid 2035804] futex(0x7ffff7cb7d00, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>
164017:[pid 2035817] futex(0x7ffff7cb7d00, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
164018:[pid 2035819] futex(0x7ffff607e640, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>
164027:[pid 2035804] futex(0x7ffff607e640, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
164034:[pid 2035804] <... accept4 resumed>0x7fffffff6c60, [110], SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)
164037:[pid 2035804] futex(0x7ffff7cb7d00, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>

```


```
 wc -l out
164235 out

rg -i "futex|unavail" out | wc -l
43453

```

some sort of threading / tokio problem?


> rbw 1.12.1

**Comment by HosseyNJF** [comment]
@doy I also managed to reproduce issue using rbw commands in chezmoi templates. My dotfiles repository is available here: https://github.com/hosseynjf/dotfiles.

In order to reproduce, comment the line in `.chezmoiscripts/darwin/run_onchange_before_install-bitwarden.sh.tmpl` where the pinentry is set to `pinentry-tty`.

**Comment by adrianschlatter** [comment]
I also observe far slower performance on macOS compared to Linux (native and in WSL). On macOS, rbw feels almost as slow as the original bitwarden cli. I can contribute some data comparing rbw on different OSes:

On Windows 11, WSL, Ubuntu 24.04.02 LTS

```zsh
> time rbw list | wc -l
486
rbw list  0.03s user 0.05s system 29% cpu 0.275 total
wc -l  0.00s user 0.00s system 0% cpu 0.275 total
```


On macOS, 15.5, 24F74:

```zsh
❯ time rbw list | wc -l
     486
rbw list  0.60s user 4.12s system 98% cpu 4.812 total
wc -l  0.00s user 0.00s system 0% cpu 4.810 total
```


On Ubuntu 24.04.02 LTS (Raspi 5):

```zsh
❯ time rbw list | wc -l
486
rbw list  0.03s user 0.12s system 57% cpu 0.263 total
wc -l  0.00s user 0.00s system 1% cpu 0.262 total
```

**Comment by calvinbui** [comment]
I can reproduce this and believe it's macOS-specific, TTY-dependent, and scales with vault size.

`rbw get <name>` decrypts every cipher in the vault to search by name, and each decryption request calls `ttyname()` via `rustix::termios::tty::_ttyname` in `get_environment`. 

On macOS, `ttyname` is implemented as a linear scan of `/dev` (`devname_r` → `readdir` + `lstat` per entry). Vault items * full `/dev` scans = seconds of kernel time. 

On Linux `ttyname` is O(1)-ish (procfs), the scan cost is macOS-only and disappears when stdin isn't a TTY, and is also proportional to vault size.

M2 MacBook Air with >2000 vault items:

```
$ time rbw get --raw 'SOPS AGE Key' >/dev/null
0.32s user 3.56s system 96% cpu 4.005 total     # by name, from a terminal

$ time rbw get --raw 'SOPS AGE Key' >/dev/null </dev/null
0.10s user 0.09s system 54% cpu 0.346 total     # by name, no TTY → ttyname fails fast

$ time rbw get --raw '167c109a-…' >/dev/null
0.01s user 0.02s system 43% cpu 0.054 total     # by UUID → no search, no per-item requests
```

Call graph from sample during a slow `rbw get`:

```
rbw::commands::find_entry
  rbw::commands::decrypt_search_cipher
    rbw::commands::decrypt_field
      rbw::actions::decrypt
        rbw::actions::get_environment
          rustix::termios::tty::_ttyname
            ttyname_r (libsystem_c)
              devname_r → lstat   ← ~99% of samples in this /dev scan
```
